File: dotnet-order-processing/Services/OrderService.cs
using DotnetOrderProcessing.Models;
using DotnetOrderProcessing.Repositories;
using System.Linq;

namespace DotnetOrderProcessing.Services;

public class OrderService : IOrderService
{
    private static readonly string[] ValidStatuses = new[] { "PENDING", "PROCESSING", "SHIPPED", "DELIVERED", "CANCELLED" };
    private readonly IOrderRepository _repo;

    public OrderService(IOrderRepository repo) => _repo = repo;

    public Order Create(OrderCreateDto dto)
    {
        ValidateItems(dto.Items);
        var order = new Order
        {
            Id = Guid.NewGuid().ToString(),
            Items = dto.Items,
            Total = dto.Items.Sum(i => i.Price * i.Quantity),
            TotalPaid = 0,
            Status = "PENDING",
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };
        _repo.Save(order);
        return order;
    }

    public Order? GetById(string id) => _repo.GetById(id);

    public IEnumerable<Order> List(string? status = null) => _repo.List(status);

    public Order UpdateStatus(string id, string status)
    {
        if (!ValidStatuses.Contains(status)) throw new ArgumentException("invalid status");
        var order = _repo.GetById(id) ?? throw new KeyNotFoundException();
        order.Status = status;
        order.UpdatedAt = DateTime.UtcNow;
        _repo.Update(order);
        return order;
    }

    public Order Cancel(string id)
    {
        var order = _repo.GetById(id) ?? throw new KeyNotFoundException();
        if (order.Status != "PENDING") throw new InvalidOperationException("only PENDING orders can be cancelled");
        if (order.TotalPaid > 0) throw new InvalidOperationException("cannot cancel order with existing payments");
        order.Status = "CANCELLED";
        order.UpdatedAt = DateTime.UtcNow;
        _repo.Update(order);
        return order;
    }

    public IEnumerable<Order> TransitionPendingToProcessing()
    {
        var pending = _repo.List("PENDING").ToList();
        var updated = new List<Order>();
        foreach (var o in pending)
        {
            if (o.TotalPaid >= o.Total)
            {
                o.Status = "PROCESSING";
                o.UpdatedAt = DateTime.UtcNow;
                _repo.Update(o);
                updated.Add(o);
            }
        }
        return updated;
    }

    public Payment AddPayment(string orderId, PaymentCreateDto dto)
    {
        var order = _repo.GetById(orderId) ?? throw new KeyNotFoundException();
        if (order.Status != "PENDING") throw new InvalidOperationException("payments can only be made for PENDING orders");
        if (dto.Amount <= 0) throw new ArgumentException("amount must be a positive number");
        
        var remaining = order.Total - order.TotalPaid;
        if (dto.Amount > remaining) throw new ArgumentException($"overpayment: cannot pay {dto.Amount}, remaining balance is {remaining}");
        
        var payment = new Payment
        {
            Id = Guid.NewGuid().ToString(),
            OrderId = orderId,
            Amount = dto.Amount,
            PaymentDate = DateTime.UtcNow,
            PaymentMethod = dto.PaymentMethod ?? "unknown"
        };
        
        // Save payment first
        var savedPayment = _repo.SavePayment(payment);
        
        // Get fresh order reference and update TotalPaid
        var updatedOrder = _repo.GetById(orderId) ?? throw new KeyNotFoundException();
        updatedOrder.TotalPaid += dto.Amount;
        updatedOrder.UpdatedAt = DateTime.UtcNow;
        _repo.Update(updatedOrder);
        
        return savedPayment;
    }

    public IEnumerable<Payment> GetPayments(string orderId)
    {
        var order = _repo.GetById(orderId) ?? throw new KeyNotFoundException();
        return _repo.GetPayments(orderId);
    }

    private void ValidateItems(List<Item> items)
    {
        if (items == null || items.Count == 0) throw new ArgumentException("items must be provided");
        foreach (var it in items)
        {
            if (string.IsNullOrEmpty(it.ProductId)) throw new ArgumentException("item.productId required");
            if (it.Quantity <= 0) throw new ArgumentException("item.quantity must be > 0");
            if (it.Price < 0) throw new ArgumentException("item.price must be >= 0");
        }
    }
}

File: nodejs-order-processing/src/services/orderService.js
const store = require('../store/inMemoryStore');
const { v4: uuidv4 } = require('uuid');
const logger = require('../utils/logger');

const VALID_STATUSES = ['PENDING', 'PROCESSING', 'SHIPPED', 'DELIVERED', 'CANCELLED'];

function computeTotal(items) {
  return items.reduce((s, it) => s + (it.price || 0) * (it.quantity || 0), 0);
}

function validateItems(items) {
  if (!Array.isArray(items) || items.length === 0) throw new Error('items must be a non-empty array');
  items.forEach(it => {
    if (!it.productId) throw new Error('item.productId required');
    if (!Number.isInteger(it.quantity) || it.quantity <= 0) throw new Error('item.quantity must be integer > 0');
    if (typeof it.price !== 'number' || it.price < 0) throw new Error('item.price must be number >= 0');
  });
}

const OrderService = {
  createOrder(data) {
    const items = data.items;
    validateItems(items);
    const total = computeTotal(items);
    const order = {
      id: uuidv4(),
      items,
      total,
      totalPaid: 0,
      status: 'PENDING',
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };
    store.save(order);
    logger.info(`Order created: ${order.id}, total: ${total}, totalPaid: 0`);
    return order;
  },

  getOrder(id) {
    return store.getById(id);
  },

  listOrders(status) {
    return store.list(status);
  },

  updateStatus(id, status) {
    if (!VALID_STATUSES.includes(status)) throw new Error('invalid status');
    const order = store.getById(id);
    if (!order) throw new Error('not found');
    order.status = status;
    order.updatedAt = new Date().toISOString();
    store.update(order);
    logger.info(`Order ${id} status updated to ${status}`);
    return order;
  },

  cancelOrder(id) {
    const order = store.getById(id);
    if (!order) throw new Error('not found');
    if (order.status !== 'PENDING') throw new Error('only PENDING orders can be cancelled');
    if (order.totalPaid > 0) throw new Error('cannot cancel order with existing payments');
    order.status = 'CANCELLED';
    order.updatedAt = new Date().toISOString();
    store.update(order);
    logger.info(`Order ${id} cancelled`);
    return order;
  },

  transitionPendingToProcessing() {
    const pending = store.list('PENDING');
    const updated = [];
    pending.forEach(order => {
      if (order.totalPaid >= order.total) {
        order.status = 'PROCESSING';
        order.updatedAt = new Date().toISOString();
        store.update(order);
        updated.push(order);
        logger.info(`Order ${order.id} transitioned to PROCESSING (paid: ${order.totalPaid}/${order.total})`);
      }
    });
    return updated;
  },

  addPayment(orderId, amount, paymentMethod = 'unknown') {
    const order = store.getById(orderId);
    if (!order) throw new Error('not found');
    if (order.status !== 'PENDING') throw new Error('payments can only be made for PENDING orders');
    if (typeof amount !== 'number' || amount <= 0) throw new Error('amount must be a positive number');
    const remaining = order.total - order.totalPaid;
    if (amount > remaining) throw new Error(`overpayment: cannot pay ${amount}, remaining balance is ${remaining}`);
    
    const payment = {
      id: uuidv4(),
      orderId,
      amount,
      paymentDate: new Date().toISOString(),
      paymentMethod,
    };
    
    const savedPayment = store.savePayment(payment);
    order.totalPaid += amount;
    order.updatedAt = new Date().toISOString();
    store.update(order);
    
    logger.info(`Payment added to order ${orderId}: amount=${amount}, totalPaid=${order.totalPaid}`);
    return savedPayment;
  },

  getPayments(orderId) {
    const order = store.getById(orderId);
    if (!order) throw new Error('not found');
    return store.getPayments(orderId);
  }
};

module.exports = OrderService;
